Add hw info for Shelly devices

chemelli74 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add hardware info for Shelly devices

Note: need an updated `aioshelly`

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

chemelli74 committed:
Add hw info for Shelly devices

thecode commented:
I am against adding this as implemented to Shelly:
- Shelly does not maintain this field, it is mostly used by them to follow up manufacturing batches (PR also refer to a batch field), most of my devices still show invalid or irrelevant information in those fields. 
- Adding information which will point users to worry why their devices is marked as "pre-production" or "dev-prototype" will just make users worried why they have such device, for example: https://github.com/home-assistant/core/issues/47947#issuecomment-913213072 user asks why he has a "dev-prototype".
- Most of Gen1 devices does not contain this field and in Gen2 it is not published how to fetch this data (marked as "batch").
- We should not implement something just because it exists in HA, for many devices this has useful information such as TP-Link devices which has completely different hardware under the same Model/PN or some Xiaomi devices which are incompatible in some hardware versions. Shelly does not use this information to use different firmware for the device, whenever the hardware changed a new device model was created, for example Shelly Button 1 & Button 2, Shelly Dimmer 1 & Dimmer 2. If you look at this devices they are identical from user perspective but due to hardware change they use a different model.
- For the past 3 years+ I am using Shelly there was not a single case I needed to know this information.
- **A better use for this field is to show the user the device generation** which does give information about the hardware, Gen1 & Gen2 has different software capabilities and the documentation clearly point you to different sections according to the device generation, while we are in details and know by the model which generation the device is, the user does not know and this could help users.

thecode reviewed (changes requested):
See my previous comment, we should consider showing device generation instead

chemelli74 committed:
Better rappresentation

chemelli74 commented:
Changed approach and added the info I'm interested in:

**Gen1:**

![image](https://user-images.githubusercontent.com/57354320/147422059-a02c52d5-8f33-424a-bb52-d75f34f7bb43.png)

**Gen2:**

![image](https://user-images.githubusercontent.com/57354320/147422088-d06a745d-6d97-418c-9cf7-e6d510f968d2.png)

Open to discuss of course ;-)

Simone

bieniu commented:
I like the new approach with gen1/gen2 information.

thecode commented:
I still think we should not try to parse the production info this is sure going to fail many times in the future, it is not displayed by the manufacture (web or app) and does not provide any useful information for the user regarding this unit compatibility with Home Assistant. The `parse_device_prod_date` looks fragile and and just by looking at 10 devices I have 4 will fail the parsing.

chemelli74 commented:
> I still think we should not try to parse the production info this is sure going to fail many times in the future, it is not displayed by the manufacture (web or app) and does not provide any useful information for the user regarding this unit compatibility with Home Assistant. The `parse_device_prod_date` looks fragile and and just by looking at 10 devices I have 4 will fail the parsing.

I'm interested in knowing when a device fails if it's a old one or not.
Please paste the json of the 4 failing. I'm interested in looking into it.

Simone

mib1185 commented:
I'm with @thecode - the used data fields are not official documented, so can be changed at any time, so we should not use them for any common (user facing) representations (like the device card).
Showing the gen version is absolutely ok and can - when available - be enriched with the technical device type (from [`/shelly`](https://shelly-api-docs.shelly.cloud/gen1/#shelly) and equivalent in gen2) - eg. "gen1 (SHPLG-S)"
For sure, the hwinfo (gen1) and the batch data (gen2) can be added as debug message (if they are available)

chemelli74 commented:
> I'm with @thecode - the used data fields are not official documented, so can be changed at any time, so we should not use them for any common (user facing) representations (like the device card). 

As I already wrote, I spoked with Allterco devs and they told me they just miss the info in doc, and this will be added.

> For sure, the hwinfo (gen1) and the batch data (gen2) can be added as debug message (if they are available)

For that to work we still need to change aioshelly ;-)

Simone

chemelli74 committed:
Remove date parsing

mib1185 commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -431,6 +431,7 @@ def async_setup(self) -> None:
             manufacturer="Shelly",
             model=aioshelly.const.MODEL_NAMES.get(self.model, self.model),
             sw_version=sw_version,
+            hw_version=f"gen{self.device.gen}",
```
```suggestion
            hw_version=f"gen{self.device.gen} ({self.model})",
```

mib1185 commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -703,6 +704,7 @@ def async_setup(self) -> None:
             manufacturer="Shelly",
             model=aioshelly.const.MODEL_NAMES.get(self.model, self.model),
             sw_version=sw_version,
+            hw_version=f"gen{self.device.gen}",
```
```suggestion
            hw_version=f"gen{self.device.gen} ({self.model})",
```

mib1185 reviewed (commented):
as discussed on discord, please add the technical model name to the hardware info

chemelli74 committed:
Added model as requested

chemelli74 commented:
> as discussed on discord, please add the technical model name to the hardware info

done as requested
